### DISCUSSION TITLE
Translation of months and days in the datetime component

### DISCUSSION THREAD

**Discussion by dcoevoet51** [formio/formio.js#4878] [General]
Hello, 

I would like to translate the months and days in the datetime component but I don't know how to do this and if it's possible.

In this example days and months are not translated: http://formio.github.io/formio.js/app/examples/language.html

I have tried this code for the datetime component but it does not translate the month:                         
                        {
                            type: 'datetime',
                            key: 'firstName',
                            label: 'First Name',
                            placeholder: 'Nom à renseigner',
                            input: true,
                            january: 'Janvier',
                            format: 'Format',
                        },

Thanks by advance, 

Best regards,

**Comment by randallknutson** [comment]
These are translatable. You can see the raw values here: https://github.com/formio/formio.js/blob/master/src/components/day/Day.js#L94-L106

**Comment by dcoevoet51** [comment]
Thank's for you reply.

However I need to translate the different months when I call the component, like in this example: http://formio.github.io/formio.js/app/examples/language.html

I cannot modify directly the file Day.js

**Comment by randallknutson** [comment]
You can use the values wrapped in t('') to translate the months. You shouldn't modify that file directly. What I was pointing out is the strings that are available to translate using the translation mechanism.

**Comment by dcoevoet51** [comment]
After trying different ways I cannot translate the months name in the datetime component.

Could you please give me a whole example of form creating with a datetime component and different labels for the months?

Thanks by advance, 

Best regards,

**Comment by TrampleFirstStrike** [comment]
Hello dcoevoet51,

Jordan here with Form.io Support Team. We appreciate your interest and use of Form.io. 

We would be happy to help you create a working example of form creation with a datetime component and different labels as you desire under a Support Package . We offer Support Packages to help you make the most of Form.io. This will give you one-on-one time with one of our Senior Developers in minimum one hour increments for training or development services.  Let me know if you would like me to provide you with pricing by emailing:

Jordan@form.io

**Comment by chaseven** [comment]
I know it is been a while since you posted but I hope this help some people in the future.

I have done a basic example on fiddle (just keep in mind this solution has the form.io 3x since that one was using flatpickr
[fiddle example](https://jsfiddle.net/dat3r2xv/)

Basically I am setting the language on the flatpickr, then calling redraw to reload the screen.

includes:
https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css
https://unpkg.com/formiojs@3.29.0/dist/formio.full.min.css
https://unpkg.com/formiojs@3.29.0/dist/formio.full.min.js
https://npmcdn.com/flatpickr@4.6.6/dist/l10n/pt.js

html:
```
<div class="page-content">

<div class="btn-group">
  <button type="button" class="btn btn-primary" onclick="setLanguage('pt')">Portugues</button>
  <button type="button" class="btn btn-primary" onclick="setLanguage('en')">English</button>
</div>

  <div class="container-fluid">
    <div id="formio"></div>
  </div>
</div>
```


Js:

```
flatpickr.localize(flatpickr.l10ns.pt);

var myform;
Formio.createForm(document.getElementById('formio'), 

{
  components: [
    {
      type: 'datetime',
      key: 'dob',
      label: 'Date of Birth',
      input: true
    },
    
    {
      type: 'button',
      action: 'submit',
      label: 'Submit',
      theme: 'primary'
    }
  ]
}, {
  language: 'pt',
  i18n: {
    pt: {
      'Date of Birth': 'Data de Nascimento',
      "Submit": "Enviar"
    }
  }
}

).then(function(form) {
	myform = form;
  // Defaults are provided as follows.
  form.submission = {
    
  };
  
  window.setLanguage = function(lang) {
  
  if(lang=='pt'){
  flatpickr.localize(flatpickr.l10ns.pt);
  }
  else {
  flatpickr.localize(flatpickr.l10ns.en);
  }
  form.redraw();
    form.language = lang;
    
  };
  
  // Register for the submit event to get the completed submission.
  form.on('submit', function(submission) {
    console.log('Submission was made!', submission);
  });
  
  // Everytime the form changes, this will fire.
  form.on('change', function(changed) {
    console.log('Form was changed', changed);
  });
});
```

cheers,

**Comment by bobslee** [comment]
I got `flatpickr` locale working for an initial loaded Form. 
The Datetime component got translated.

However, why doesn't the `setLanguage` (below) work like expected?
(also explained above)

Also I got an error: `flatpickr invalid locale nl`
Therefor I needed to import the `flatpickr` **locales** but also the **libary** (see below). 
Why these ain't included in `formio.full.min.js`?

**So basically I added / imported...**

```javascript
<script src="/formio/static/installed/4.13.2/formio.full.min.js"</script>
<script src="https://cdn.jsdelivr.net/npm/flatpickr"></script>
<script src="https://npmcdn.com/flatpickr/dist/l10n/nl.js"></script>
<script src="https://npmcdn.com/flatpickr/dist/l10n/pt.js"></script>
<script src="https://npmcdn.com/flatpickr/dist/l10n/default.js"></script>
```

**Problem below - the `setLanguage` doesn't update the Datetime component (flatpickr)**

```javascript
Formio.createForm(document.getElementById('formio_form'), self.schema, self.options).then(function(form) {
    // Language
    if ('language' in self.options) {
        form.language = self.options['language'];
    }
    window.setLanguage = function(lang) {
        // https://github.com/formio/formio.js/issues/1014
        // flatpickr.localize(flatpickr.l10ns[lang]);
        flatpickr.localize(flatpickr.l10ns.pt);
        form.redraw();
        form.language = lang;
    };
   .... // events etc
```

**Comment by bobslee** [comment]
@chaseven, @randallknutson any clue or suggestion why my example doesn't work yet?

**Comment by chaseven** [comment]
Hi @bobslee , sorry it took me so long to reply. Before we start just want to let you know that I am not a formIO specialist, so maybe my solution is not the best option but worked fine for me.

First, regarding the flatpickr location I know it is "frustrating" having to add this but I got say it would suck if it were any other way. Imagine including by default 100 languages and not even using any of that because your application runs only on one language? That would be a waste of resource, so I am glad they didn't add that into the min.js 

Second, on my tests one trick thing was the order of the includes, since the flatpickr is included inside the formio.min.js you need  to add the language include after that otherwise you will get a flatpickr not defined. From the look iof your code you are doing this correctly but I just want to emphasise this in case your includes are in a different order on your final project.

Third, the script you posted looks quite different form my example on fiddle. On your code you are setting the initial locale inside of your "create", however on my example my initial locale is defined outside my create


```
flatpickr.localize(flatpickr.l10ns.pt);
//then:
Formio.createForm(document.getElementById('formio') ,....
```
if you set the locale after the form is created is already too late (unless you refresh or redraw, but it is a waste of resource doing so, since you can simple set before creating). By doing that when formio api runs it will build the components using the correct/selected language.

On my code the only reason I put that redraw and the function "window.setLanguage = function(lang)" inside the create is because I was testing the buttons to change the language on the fly, but on my final project I am not doing that because I also have other elements on my UI that are not part of formIO that also needs to be reloaded with the new language.

Additionally your "if ('language' in self.options)" is quite redundant, from my tests if you simple provide the language on the create constructor the api will handle everything

here is an example of the "options" parameter including the default keys from formIO (version 3.x)

````
{
    "language": "es",
    "i18n": {
        "es": {
            "complete": "Envío completo",
            "error": "Se necesita corregir los siguientes errores antes de proceder",
            "required": "{{field}} es necesario",
            "pattern": "{{field}} no satisface el modelo {{pattern}}",
            "minLength": "{{field}} debe de contener mas de {{length}} caracteres.",
            "maxLength": "{{field}} debe de contener menos de {{length}} caracteres.",
            "minWords": "{{field}} debe de tener mas de {{length}} palabras.",
            "maxWords": "{{field}} debe de tener menos de {{length}} palabras.",
            "min": "{{field}} no puede ser menor a {{min}}.",
            "max": "{{field}} no puede ser mayor a {{max}}.",
            "minSelectedCount": "Debes de seleccionar por lo menos {{minCount}} items para continuar.",
            "maxSelectedCount": "Solo puedes seleccionar hasta {{maxCount}} items para continuar.",
            "maxDate": "{{field}} no puede ser mayor a {{- maxDate}}",
            "minDate": "{{field}} no puede ser menor a {{- minDate}}",
            "invalid_email": "{{field}} debe ser tener una dirección de email válida.",
            "invalid_url": "{{field}} debe de tener un url válido.",
            "invalid_regex": "{{field}} no satisface el modelo {{regex}}.",
            "invalid_date": "{{field}} fecha no válida.",
            "invalid_day": "{{field}} dia del mes invalido.",
            "mask": "{{field}} no satisface el modelo.",
            "stripe": "{{stripe}}",
            "month": "Mes",
            "day": "Día",
            "year": "Año",
            "january": "Enero",
            "february": "Febrero",
            "march": "Marzo",
            "april": "Abril",
            "may": "Mayo",
            "june": "Junio",
            "july": "Julio",
            "august": "Agosto",
            "september": "Septiembre",
            "october": "Octubre",
            "november": "Noviembre",
            "december": "Diciembre",
            "next": "Siguiente",
            "previous": "Anterior",
            "cancel": "Cancelar",
            "Save as Draft": "Guardar como borrador",
            "submit": "Enviar",
            "Loading...": "Cargando...",
            "No results found": "No se encontraron registros",
            "No choices to choose from": "No hay opciones para elegir",
            "Only unique values can be added": "Solo puedes anadir valores unicos",
            "Only values matching specific conditions can be added": "Solo puedes anadir valores que coincidan con ciertas condiciones",
            "Type to search": "Escriba para busqueda",
            "Press to select": "Presiona para seleccionar"
        }
    }
}
````


I hope this all helps you.

Cheers,
Michael

**Comment by experiment0** [comment]
Hello

For datetime field Formio use Flatpickr.

Flatpickr options added to field `customOptions`.

Example of Flatpickr localization options https://flatpickr.js.org/localization/#using-modules 

Example form scheme with localisation datetime field:

```
const Russian = {
  weekdays: {
    shorthand: ['Вс', 'Пн', 'Вт', 'Ср', 'Чт', 'Пт', 'Сб'],
    longhand: ['Воскресенье', 'Понедельник', 'Вторник', 'Среда', 'Четверг', 'Пятница', 'Суббота'],
  },
  months: {
    shorthand: [
      'Янв', 'Фев', 'Март', 'Апр', 'Май', 'Июнь', 'Июль', 'Авг', 'Сен', 'Окт', 'Ноя', 'Дек',
    ],
    longhand: [
      'Январь', 'Февраль', 'Март', 'Апрель', 'Май', 'Июнь', 'Июль', 'Август', 'Сентябрь', 'Октябрь', 'Ноябрь', 'Декабрь',
    ],
  },
  firstDayOfWeek: 1,
  ordinal: () => '',
  rangeSeparator: ' — ',
  weekAbbreviation: 'Нед.',
  scrollTitle: 'Прокрутите для увеличения',
  toggleTitle: 'Нажмите для переключения',
  amPM: ['ДП', 'ПП'],
  yearAriaLabel: 'Год',
  time_24hr: true,
}

const formScheme = {
  "display": "form",
  "components": [
    {
      "label": "Date / Time",
      "format": "yyyy-MM-dd",
      "tableView": false,
      "enableMinDateInput": false,
      "datePicker": {
        "disableWeekends": false,
        "disableWeekdays": false
      },
      "enableMaxDateInput": false,
      "enableTime": false,
      "customOptions": {
        "locale": Russian,
      },
      "key": "dateTime",
      "type": "datetime",
      "input": true,
      "widget": {
        "type": "calendar",
        "displayInTimezone": "viewer",
        "locale": "ru",
        "useLocaleSettings": false,
        "allowInput": true,
        "mode": "single",
        "enableTime": false,
        "noCalendar": false,
        "format": "yyyy-MM-dd",
        "hourIncrement": 1,
        "minuteIncrement": 1,
        "time_24hr": false,
        "minDate": null,
        "disableWeekends": false,
        "disableWeekdays": false,
        "maxDate": null,
      }
    },
    {
      "label": "Submit",
      "showValidations": false,
      "tableView": false,
      "key": "submit",
      "type": "button",
      "input": true,
    }
  ]
}
```

**Comment by AlexByte** [comment]
Why not to use this way?
```
<script src="https://cdn.form.io/flatpickr/flatpickr.min.js"></script>
<link rel="stylesheet" href="https://cdn.form.io/flatpickr/flatpickr.min.css">
<script src="https://npmcdn.com/flatpickr/dist/l10n/ru.js"></script> <-- localization file -->

```
